<a target="_blank" href="https://colab.research.google.com/github/lukebarousse/Int_SQL_Data_Analytics_Course/blob/main/8_Project/1_Final_Project.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# Final Project

In [2]:
import sys
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

# If running in Google Colab, install PostgreSQL and restore the database
if 'google.colab' in sys.modules:
    # Install PostgreSQL
    !sudo apt-get install postgresql -qq > /dev/null 2>&1

    # Start PostgreSQL service (suppress output)
    !sudo service postgresql start > /dev/null 2>&1

    # Set password for the 'postgres' user to avoid authentication errors (suppress output)
    !sudo -u postgres psql -c "ALTER USER postgres WITH PASSWORD 'password';" > /dev/null 2>&1

    # Create the 'colab_db' database (suppress output)
    !sudo -u postgres psql -c "CREATE DATABASE contoso_100k;" > /dev/null 2>&1

    # Download the PostgreSQL .sql dump
    !wget -q -O contoso_100k.sql https://github.com/lukebarousse/Int_SQL_Data_Analytics_Course/releases/download/v.0.0.0/contoso_100k.sql

    # Restore the dump file into the PostgreSQL database (suppress output)
    !sudo -u postgres psql contoso_100k < contoso_100k.sql > /dev/null 2>&1

    # Shift libraries from ipython-sql to jupysql
    !pip uninstall -y ipython-sql > /dev/null 2>&1
    !pip install jupysql > /dev/null 2>&1

# Load the sql extension for SQL magic
%load_ext sql

# Connect to the PostgreSQL database
%sql postgresql://postgres:password@localhost:5432/contoso_100k

# Enable automatic conversion of SQL results to pandas DataFrames
%config SqlMagic.autopandas = True

# Disable named parameters for SQL magic
%config SqlMagic.named_parameters = "disabled"

# Display pandas number to two decimal places
pd.options.display.float_format = '{:.2f}'.format

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Connecting and switching to connection 'postgresql://postgres:***@localhost:5432/contoso_100k'

## Background

You're a **data analyst at an e-commerce company**. Your stakeholders on marketing & finance teams need insights to improve customer retention and maximize revenue. They have three key questions:

1️⃣ **Who are our most valuable customers?** (Customer Segmentation)

2️⃣ **How do different customer groups generate long-term revenue?** (Cohort-Based LTV) 

3️⃣ **Which customers haven’t purchased recently?** (Retention Analysis)

Your job is to create a structured analysis using SQL that answers these questions and provides actionable insights for the business.

## Analysis

### 0️⃣ Data Cleaning & Preprocessing

#### Overview

- Before starting, create a **cleaned view** to ensure data consistency.
- Standardize customer names, country, age, the cohort, first purchase date, and total net revenue.

#### Query Steps

1. Take same query that's in the `cohort_analysis` and add in `MIN(orderdate) AS first_purchase_date` to the `cohort` CTE.

In [3]:
%%sql

WITH cohort AS (
    SELECT 
        customerkey,
        EXTRACT(year FROM MIN(orderdate)) AS cohort_year,
        MIN(orderdate) AS first_purchase_date
    FROM sales
    GROUP BY sales.customerkey
        )
        
SELECT s.customerkey,
    c.cohort_year,
    s.orderdate,
    c.first_purchase_date,
    sum(s.quantity::double precision * s.netprice * s.exchangerate) AS total_net_revenue
FROM sales s
    LEFT JOIN cohort c ON c.customerkey = s.customerkey
GROUP BY 
    s.customerkey, 
    c.cohort_year, 
    s.orderdate, 
    c.first_purchase_date;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

83099 rows affected.

customerkey cohort_year   orderdate first_purchase_date  \
0               15        2021  2021-03-08          2021-03-08   
1              180        2018  2018-07-28          2018-07-28   
2              180        2018  2023-08-28          2018-07-28   
3              185        2019  2019-06-01          2019-06-01   
4              243        2016  2016-05-19          2016-05-19   
...            ...         ...         ...                 ...   
83094      2099697        2022  2022-09-13          2022-09-13   
83095      2099711        2016  2016-08-13          2016-08-13   
83096      2099711        2016  2017-08-14          2016-08-13   
83097      2099743        2022  2022-03-17          2022-03-17   
83098      2099743        2022  2023-02-11          2022-03-17   

       total_net_revenue  
0                2217.41  
1                 525.31  
2                1984.90  
3                1395.52  
4                 287.67  
...                  ...  
83094              38.20  
83095            2067.75  
83096            3940.92  
83097             469.62  
83098             598.46  

[83099 rows x 5 columns]

2. Put the main query into a CTE named `cohort_data` and add an alias to `cohort_data` named `cd` and select individually every column. 

In [4]:
%%sql

WITH cohort AS (
    SELECT 
        customerkey,
        EXTRACT(year FROM MIN(orderdate)) AS cohort_year,
        MIN(orderdate) AS first_purchase_date
    FROM sales
    GROUP BY sales.customerkey
),

-- Put query into a CTE
cohort_data AS (
	SELECT 
		s.customerkey,
		c.cohort_year,
		s.orderdate,
		c.first_purchase_date,
    	sum(s.quantity::double precision * s.netprice * s.exchangerate) AS total_net_revenue
	FROM sales s
		LEFT JOIN cohort c ON c.customerkey = s.customerkey
	GROUP BY 
		s.customerkey, 
		c.cohort_year, 
		s.orderdate, 
		c.first_purchase_date
) 

-- Added
SELECT
	cd.customerkey,
	cd.cohort_year,
	cd.first_purchase_date,
	cd.orderdate,
	cd.total_net_revenue
FROM cohort_data cd 

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

83099 rows affected.

customerkey cohort_year first_purchase_date   orderdate  \
0               15        2021          2021-03-08  2021-03-08   
1              180        2018          2018-07-28  2018-07-28   
2              180        2018          2018-07-28  2023-08-28   
3              185        2019          2019-06-01  2019-06-01   
4              243        2016          2016-05-19  2016-05-19   
...            ...         ...                 ...         ...   
83094      2099697        2022          2022-09-13  2022-09-13   
83095      2099711        2016          2016-08-13  2016-08-13   
83096      2099711        2016          2016-08-13  2017-08-14   
83097      2099743        2022          2022-03-17  2022-03-17   
83098      2099743        2022          2022-03-17  2023-02-11   

       total_net_revenue  
0                2217.41  
1                 525.31  
2                1984.90  
3                1395.52  
4                 287.67  
...                  ...  
83094              38.20  
83095            2067.75  
83096            3940.92  
83097             469.62  
83098             598.46  

[83099 rows x 5 columns]

3. LEFT JOIN `customer` table on the `customerkey` to get the customer's `givenname` and `surname` and concatenate the names. Also get the customer's `age` and `countryfull`.  

In [5]:
%%sql

WITH cohort AS (
    SELECT 
        customerkey,
        EXTRACT(year FROM MIN(orderdate)) AS cohort_year,
        MIN(orderdate) AS first_purchase_date
    FROM sales
    GROUP BY sales.customerkey
),

-- Put query into a CTE
cohort_data AS (
	SELECT 
		s.customerkey,
		c.cohort_year,
		s.orderdate,
		c.first_purchase_date,
    	sum(s.quantity::double precision * s.netprice * s.exchangerate) AS total_net_revenue
	FROM sales s
		LEFT JOIN cohort c ON c.customerkey = s.customerkey
	GROUP BY 
		s.customerkey, 
		c.cohort_year, 
		s.orderdate, 
		c.first_purchase_date
)

SELECT
	cd.customerkey,
	CONCAT(TRIM(c.givenname), ' ', TRIM(c.surname)) AS customer_name, -- Added
    c.countryfull, -- Added
    c.age, -- Added
	cd.cohort_year,
	cd.first_purchase_date,
	cd.orderdate,
	cd.total_net_revenue
FROM cohort_data cd 
LEFT JOIN customer c ON c.customerkey = cd.customerkey -- Added
;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

83099 rows affected.

customerkey        customer_name    countryfull  age cohort_year  \
0               15      Julian McGuigan      Australia   55        2021   
1              180    Gabriel Bosanquet      Australia   65        2018   
2              180    Gabriel Bosanquet      Australia   65        2018   
3              185   Gabrielle Castella      Australia   40        2019   
4              243        Maya Atherton      Australia   66        2016   
...            ...                  ...            ...  ...         ...   
83094      2099697       Phillipp Maier  United States   54        2022   
83095      2099711  Katerina Pavlícková  United States   80        2016   
83096      2099711  Katerina Pavlícková  United States   80        2016   
83097      2099743      Luciana Almonte  United States   21        2022   
83098      2099743      Luciana Almonte  United States   21        2022   

      first_purchase_date   orderdate  total_net_revenue  
0              2021-03-08  2021-03-08            2217.41  
1              2018-07-28  2018-07-28             525.31  
2              2018-07-28  2023-08-28            1984.90  
3              2019-06-01  2019-06-01            1395.52  
4              2016-05-19  2016-05-19             287.67  
...                   ...         ...                ...  
83094          2022-09-13  2022-09-13              38.20  
83095          2016-08-13  2016-08-13            2067.75  
83096          2016-08-13  2017-08-14            3940.92  
83097          2022-03-17  2022-03-17             469.62  
83098          2022-03-17  2023-02-11             598.46  

[83099 rows x 8 columns]

4. Create a view called `cleaned_customer`.

In [6]:
%%sql

--CREATE VIEW cleaned_customer AS -- commented out to avoid overwriting
WITH cohort AS (
    SELECT 
        customerkey,
        EXTRACT(year FROM MIN(orderdate)) AS cohort_year,
        MIN(orderdate) AS first_purchase_date
    FROM sales
    GROUP BY sales.customerkey
),

-- Put query into a CTE
cohort_data AS (
	SELECT 
		s.customerkey,
		c.cohort_year,
		s.orderdate,
		c.first_purchase_date,
    	sum(s.quantity::double precision * s.netprice * s.exchangerate) AS total_net_revenue
	FROM sales s
		LEFT JOIN cohort c ON c.customerkey = s.customerkey
	GROUP BY 
		s.customerkey, 
		c.cohort_year, 
		s.orderdate, 
		c.first_purchase_date
)

SELECT
	cd.customerkey,
	CONCAT(TRIM(c.givenname), ' ', TRIM(c.surname)) AS customer_name, -- Added
    c.countryfull, -- Added
    c.age, -- Added
	cd.cohort_year,
	cd.first_purchase_date,
	cd.orderdate,
	cd.total_net_revenue
FROM cohort_data cd 
LEFT JOIN customer c ON c.customerkey = cd.customerkey -- Added
;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

83099 rows affected.

customerkey        customer_name    countryfull  age cohort_year  \
0               15      Julian McGuigan      Australia   55        2021   
1              180    Gabriel Bosanquet      Australia   65        2018   
2              180    Gabriel Bosanquet      Australia   65        2018   
3              185   Gabrielle Castella      Australia   40        2019   
4              243        Maya Atherton      Australia   66        2016   
...            ...                  ...            ...  ...         ...   
83094      2099697       Phillipp Maier  United States   54        2022   
83095      2099711  Katerina Pavlícková  United States   80        2016   
83096      2099711  Katerina Pavlícková  United States   80        2016   
83097      2099743      Luciana Almonte  United States   21        2022   
83098      2099743      Luciana Almonte  United States   21        2022   

      first_purchase_date   orderdate  total_net_revenue  
0              2021-03-08  2021-03-08            2217.41  
1              2018-07-28  2018-07-28             525.31  
2              2018-07-28  2023-08-28            1984.90  
3              2019-06-01  2019-06-01            1395.52  
4              2016-05-19  2016-05-19             287.67  
...                   ...         ...                ...  
83094          2022-09-13  2022-09-13              38.20  
83095          2016-08-13  2016-08-13            2067.75  
83096          2016-08-13  2017-08-14            3940.92  
83097          2022-03-17  2022-03-17             469.62  
83098          2022-03-17  2023-02-11             598.46  

[83099 rows x 8 columns]

### 1️⃣ Customer Segmentation (Who Are Our Most Valuable Customers?)

#### Overview
- Categorize customers based on their total lifetime value (LTV).
- Assign customers to **High, Mid, and Low-value** groups using CASE WHEN.

💼 **Example Use Cases:** Enables targeted marketing and personalized experiences
- Provide VIP benefits to high-value customers (early access, premium service)
- Create targeted upgrade paths for mid-value customers through personalized promotions
- Design re-engagement campaigns for low-value customers to increase purchase frequency
- Optimize marketing spend based on customer segment potential

#### Query Steps

1. Get the customer's lifetime value (LTV). 

In [7]:
%%sql 

SELECT
    customerkey,
    SUM(total_net_revenue) AS total_ltv
FROM cleaned_customer
GROUP BY customerkey

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

49487 rows affected.

customerkey  total_ltv
0               15    2217.41
1              180    2510.22
2              185    1395.52
3              243     287.67
4              387    4655.84
...            ...        ...
49482      2099619    6709.94
49483      2099656   10404.68
49484      2099697      38.20
49485      2099711    6008.67
49486      2099743    1068.08

[49487 rows x 2 columns]

2. Get the 25th and 75th percentile of the LTV. This will help us segement the customer's (similar to the notebook [3_Advanced_Segementation.ipynb](../1_Pivot_With_Case_Statements/3_Advanced_Segmentation.ipynb)).
    - High-Value: Customers in the top 25% (75th percentile and above)
    - Mid-Value: Customers in the middle 50% (25th to 75th percentile)
    - Low-Value: Customers in the bottom 25% (below the 25th percentile)

In [10]:
%%sql 

-- Put previous main query into a CTE
WITH customer_ltv AS (
    SELECT
        customerkey,
        SUM(total_net_revenue) AS total_ltv
    FROM cleaned_customer
    GROUP BY customerkey
)

SELECT
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY total_ltv) AS percentile_25th,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY total_ltv) AS percentile_75th
FROM customer_ltv;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

1 rows affected.

percentile_25th  percentile_75th
0           843.59          5584.04

3. Using the 25th and 75th percentile, we can now segment the customers into High, Mid, and Low-value segments.

In [11]:
%%sql

WITH customer_ltv AS (
    SELECT
        customerkey,
        SUM(total_net_revenue) AS total_ltv
    FROM cleaned_customer
    GROUP BY customerkey
),

-- Put previous main query into a CTE
customer_segments AS (
    SELECT
        PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY total_ltv) AS percentile_25th,
        PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY total_ltv) AS percentile_75th
    FROM customer_ltv
)

-- Add the segments to the main query
SELECT
    c.customerkey,
    c.total_ltv,
    CASE
        WHEN c.total_ltv > percentile_75th THEN 'High-Value'
        WHEN c.total_ltv BETWEEN percentile_25th AND percentile_75th THEN 'Mid-Value'
        ELSE 'Low-Value'
    END AS customer_segment
FROM customer_ltv c,
    customer_segments cs;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

49487 rows affected.

customerkey  total_ltv customer_segment
0               15    2217.41        Mid-Value
1              180    2510.22        Mid-Value
2              185    1395.52        Mid-Value
3              243     287.67        Low-Value
4              387    4655.84        Mid-Value
...            ...        ...              ...
49482      2099619    6709.94       High-Value
49483      2099656   10404.68       High-Value
49484      2099697      38.20        Low-Value
49485      2099711    6008.67       High-Value
49486      2099743    1068.08        Mid-Value

[49487 rows x 3 columns]

4. Get the total revenue for each customer segment.

In [14]:
%%sql

WITH customer_ltv AS (
    SELECT
        customerkey,
        SUM(total_net_revenue) AS total_ltv
    FROM cleaned_customer
    GROUP BY customerkey
),

customer_segments AS (
    SELECT
        PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY total_ltv) AS percentile_25th,
        PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY total_ltv) AS percentile_75th
    FROM customer_ltv
),

-- Put previous main query into a CTE
segement_values AS (
    SELECT
        c.customerkey,
        c.total_ltv,
        CASE
            WHEN c.total_ltv > percentile_75th THEN 'High-Value'
            WHEN c.total_ltv BETWEEN percentile_25th AND percentile_75th THEN 'Mid-Value'
            ELSE 'Low-Value'
        END AS customer_segment
    FROM customer_ltv c,
    customer_segments cs
)

SELECT
    customer_segment,
    SUM(total_ltv) AS total_ltv,
    COUNT(customerkey) AS customer_count
FROM segement_values
GROUP BY customer_segment
ORDER BY total_ltv DESC
;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

3 rows affected.

customer_segment    total_ltv  customer_count
0       High-Value 135429277.27           12372
1        Mid-Value  66636451.79           24743
2        Low-Value   4341809.53           12372

#### 📊 Key Findings

- High-value segment (25% of customers) drives 66% of revenue ($135.4M)
    - 12,372 customers (25% of 49,487 total customers)
    - $135.4M / $206.4M total revenue = 66%
- Mid-value segment (50% of customers) generates 32% of revenue ($66.6M)
    - 24,743 customers (50% of 49,487 total customers)
    - $66.6M / $206.4M total revenue = 32%
- Low-value segment (25% of customers) accounts for 2% of revenue ($4.3M)
    - 12,372 customers (25% of 49,487 total customers)
    - $4.3M / $206.4M total revenue = 2%

### 💡 Business Insights

- High-Value (66% revenue):
    - Offer premium membership program to 12,372 VIP customers
    - Provide early access to new products and dedicated support
    - Focus on retention as losing one customer impacts revenue significantly
- Mid-Value (32% revenue):
    - Create upgrade paths for 24,743 customers through personalized promotions
    - Target with "next best product" recommendations based on high-value patterns
    - Potential $66.6M → $135.4M revenue opportunity if upgraded to high-value
- Low-Value (2% revenue):
    - Design re-engagement campaigns for 12,372 customers to increase purchase frequency
    - Test price-sensitive promotions to encourage more frequent purchases
    - Focus on converting $4.3M segment to mid-value through targeted offers

<img src="../8_Project/2.1_customer_segementation.png" alt="Customer Segementation by LTV" style="width: 70%; height: auto;">

### 2️⃣ Cohort-Based LTV (How Do Customer Groups Generate Long-Term Revenue?)

#### Overview
- Track **cumulative revenue per customer cohort** over time.
- Use **window functions** to calculate lifetime value trends.

💼 **Example Use Cases:** Prevents customer churn through timely intervention
- Launch personalized win-back campaigns based on past purchase behavior
- Proactively engage high-value customers showing declining activity
- Create time-sensitive offers for customers approaching churn threshold
- Use insights to improve product offerings and customer experience

#### Query Steps

1. Get the last query from the [2_Basic_Optimization.ipynb](../7_Basic_Query_Optimization/2_Basic_Optimization.ipynb) file and in the `cohort_summary` replace the `cohort_analysis` view with the `cleaned_customer` view we created for this project. 

In [ ]:
%%sql

WITH cohort_summary AS (
    SELECT
        cohort_year,
        DATE_TRUNC('month', orderdate)::date AS year_month, 
        SUM(total_net_revenue) AS total_revenue
    FROM cleaned_customer
    GROUP BY cohort_year, year_month
),

rolling_ltv AS (
    SELECT
        cohort_year,
        year_month,
        SUM(total_revenue) OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month 
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        ) AS cumulative_revenue,
        DENSE_RANK() OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month
        ) AS months_since_start,
        SUM(total_revenue) OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month
            ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
        ) AS rolling_3_month_revenue,
        COUNT(total_revenue) OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month 
            ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
        ) AS rolling_3_month_num_months
    FROM cohort_summary
)

SELECT
    cohort_year,
    year_month,
    cumulative_revenue,
    cumulative_revenue / months_since_start AS cumulative_avg_ltv,
    rolling_3_month_revenue / rolling_3_month_num_months AS rolling_3_month_avg_ltv
FROM rolling_ltv
ORDER BY 
    cohort_year, 
    year_month;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

580 rows affected.

cohort_year  year_month  cumulative_revenue  rolling_avg_ltv  \
0          2015  2015-01-01           384092.66        384092.66   
1          2015  2015-02-01          1090466.78        545233.39   
2          2015  2015-03-01          1423428.37        474476.12   
3          2015  2015-04-01          1584195.37        396048.84   
4          2015  2015-05-01          2132828.00        426565.60   
..          ...         ...                 ...              ...   
575        2023  2024-04-01         14979328.33        936208.02   
576        2024  2024-01-01           870022.73        870022.73   
577        2024  2024-02-01          2147261.90       1073630.95   
578        2024  2024-03-01          2713002.63        904334.21   
579        2024  2024-04-01          2856649.33        714162.33   

     rolling_3_month_avg_ltv  
0                  384092.66  
1                  545233.39  
2                  474476.12  
3                  400034.24  
4                  347453.74  
..                       ...  
575                141293.21  
576                870022.73  
577               1073630.95  
578                904334.21  
579                662208.87  

[580 rows x 5 columns]

2. Remove the following columns in `rolling_ltv` CTE:`rolling_3_month_revenue`, `rolling_3_month_num_months`, `months_since_start`. Remove the columns in the main query: `months_since_start`, and `rolling_3_month_avg_ltv`.

In [24]:
%%sql

WITH cohort_summary AS (
    SELECT
        cohort_year,
        DATE_TRUNC('month', orderdate)::date AS year_month, 
        SUM(total_net_revenue) AS total_revenue
    FROM cleaned_customer
    GROUP BY cohort_year, year_month
),

rolling_ltv AS (
    SELECT
        cohort_year,
        year_month,
        SUM(total_revenue) OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month 
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        ) AS cumulative_revenue,
        DENSE_RANK() OVER (
            PARTITION BY cohort_year 
            ORDER BY year_month
        ) AS months_since_start
    FROM cohort_summary
)

SELECT
    cohort_year,
    year_month,
    cumulative_revenue,
    cumulative_revenue / months_since_start AS rolling_avg_ltv
FROM rolling_ltv
ORDER BY 
    cohort_year, 
    year_month;

Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

580 rows affected.

cohort_year  year_month  cumulative_revenue  rolling_avg_ltv
0          2015  2015-01-01           384092.66        384092.66
1          2015  2015-02-01          1090466.78        545233.39
2          2015  2015-03-01          1423428.37        474476.12
3          2015  2015-04-01          1584195.37        396048.84
4          2015  2015-05-01          2132828.00        426565.60
..          ...         ...                 ...              ...
575        2023  2024-04-01         14979328.33        936208.02
576        2024  2024-01-01           870022.73        870022.73
577        2024  2024-02-01          2147261.90       1073630.95
578        2024  2024-03-01          2713002.63        904334.21
579        2024  2024-04-01          2856649.33        714162.33

[580 rows x 4 columns]

**💡 Why are we redoing `months_since_start`?**  

- **Fixes Missing Months** – `DENSE_RANK()` counted rows, but skipped months when no sales occurred. The new method ensures each month is counted accurately.  
- **Uses Actual Time Elapsed** – `DATE_PART('month', age(...))` calculates real months since first purchase, rather than just ranking available data points.  
- **More Accurate Rolling LTV** – Prevents overestimating LTV by correctly dividing by the actual number of months since the cohort started.  
- **Better for Plotting** – Cohorts are now properly aligned on a time scale, so early LTV trends aren’t distorted by missing months.  
- **Ensures Cohort Comparability** – Each cohort’s trajectory reflects real-world time instead of gaps in sales data.

3. Get the first purchase date per cohort (we're going to need this to get the months since the first purchase).
    - Create a new CTE, `cohort_first_purchase` to get the first purchase date per cohort. 
    - In `rolling_ltv` JOIN `cohort_summary` with `cohort_first_purchase`.
    - In the main query temporarily remove `cumulative_revenue / NULLIF(months_since_start, 0) AS rolling_avg_ltv` since we no longer have `months_since_start` and it would return an error.

In [25]:
%%sql

WITH cohort_summary AS (
    SELECT
        cc.cohort_year,
        DATE_TRUNC('month', cc.orderdate)::date AS year_month, 
        SUM(cc.total_net_revenue) AS total_revenue
    FROM cleaned_customer cc
    GROUP BY cc.cohort_year, year_month
),

-- Added 
cohort_first_purchase AS (
    SELECT 
        cohort_year,
        MIN(first_purchase_date) AS cohort_first_purchase_date
    FROM cleaned_customer
    GROUP BY cohort_year
),

rolling_ltv AS (
    SELECT
        cs.cohort_year,
        cs.year_month,
        cfp.cohort_first_purchase_date, -- Added 

        SUM(cs.total_revenue) OVER (
            PARTITION BY cs.cohort_year 
            ORDER BY cs.year_month 
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        ) AS cumulative_revenue
    FROM cohort_summary cs
    JOIN cohort_first_purchase cfp ON cs.cohort_year = cfp.cohort_year -- Added 
)

SELECT
    cohort_year,
    year_month,
    cohort_first_purchase_date,
    cumulative_revenue
    --cumulative_revenue / NULLIF(months_since_start, 0) AS rolling_avg_ltv
FROM rolling_ltv
ORDER BY 
    cohort_year, 
    year_month;


Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

580 rows affected.

cohort_year  year_month cohort_first_purchase_date  cumulative_revenue
0          2015  2015-01-01                 2015-01-01           384092.66
1          2015  2015-02-01                 2015-01-01          1090466.78
2          2015  2015-03-01                 2015-01-01          1423428.37
3          2015  2015-04-01                 2015-01-01          1584195.37
4          2015  2015-05-01                 2015-01-01          2132828.00
..          ...         ...                        ...                 ...
575        2023  2024-04-01                 2023-01-01         14979328.33
576        2024  2024-01-01                 2024-01-01           870022.73
577        2024  2024-02-01                 2024-01-01          2147261.90
578        2024  2024-03-01                 2024-01-01          2713002.63
579        2024  2024-04-01                 2024-01-01          2856649.33

[580 rows x 4 columns]

4. Calculate the months since first purchase in `rolling_ltv` and then use that to calculate the `cumulative_avg_ltv` in the main query. 
    - How new `months_since_start` works:
        - `age(cs.year_month, cfp.cohort_first_purchase_date)` calculates the time difference as an interval (e.g., "1 year 3 months").  
        - `DATE_PART('month', age(...))` extracts the months, and `DATE_PART('year', age(...)) * 12` converts years to months.  
        - Adding both values gives total months elapsed, and `+1` ensures the first month is counted as 1 instead of 0.  
        - This method accurately tracks time even if some months have missing data, unlike `DENSE_RANK()`.  

In [28]:
%%sql

WITH cohort_summary AS (
    SELECT
        cc.cohort_year,
        DATE_TRUNC('month', cc.orderdate)::date AS year_month, 
        SUM(cc.total_net_revenue) AS total_revenue
    FROM cleaned_customer cc
    GROUP BY cc.cohort_year, year_month
),

-- Added 
cohort_first_purchase AS (
    SELECT 
        cohort_year,
        MIN(first_purchase_date) AS cohort_first_purchase_date
    FROM cleaned_customer
    GROUP BY cohort_year
),

rolling_ltv AS (
    SELECT
        cs.cohort_year,
        cs.year_month,
        cfp.cohort_first_purchase_date,

        SUM(cs.total_revenue) OVER (
            PARTITION BY cs.cohort_year 
            ORDER BY cs.year_month 
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        ) AS cumulative_revenue,
        -- Added 
        DATE_PART('month', age(cs.year_month, cfp.cohort_first_purchase_date)) 
        + DATE_PART('year', age(cs.year_month, cfp.cohort_first_purchase_date)) * 12 
        + 1 AS months_since_start
    FROM cohort_summary cs
    JOIN cohort_first_purchase cfp ON cs.cohort_year = cfp.cohort_year
)

SELECT
    cohort_year,
    year_month,
    cohort_first_purchase_date,
    cumulative_revenue,
    cumulative_revenue / NULLIF(months_since_start, 0) AS cumulative_avg_ltv -- Added back in 
FROM rolling_ltv
ORDER BY 
    cohort_year, 
    year_month;


Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

580 rows affected.

cohort_year  year_month cohort_first_purchase_date  cumulative_revenue  \
0          2015  2015-01-01                 2015-01-01           384092.66   
1          2015  2015-02-01                 2015-01-01          1090466.78   
2          2015  2015-03-01                 2015-01-01          1423428.37   
3          2015  2015-04-01                 2015-01-01          1584195.37   
4          2015  2015-05-01                 2015-01-01          2132828.00   
..          ...         ...                        ...                 ...   
575        2023  2024-04-01                 2023-01-01         14979328.33   
576        2024  2024-01-01                 2024-01-01           870022.73   
577        2024  2024-02-01                 2024-01-01          2147261.90   
578        2024  2024-03-01                 2024-01-01          2713002.63   
579        2024  2024-04-01                 2024-01-01          2856649.33   

     rolling_avg_ltv  
0          384092.66  
1          545233.39  
2          474476.12  
3          396048.84  
4          426565.60  
..               ...  
575        936208.02  
576        870022.73  
577       1073630.95  
578        904334.21  
579        714162.33  

[580 rows x 5 columns]

5. Clean up the query and remove `cohort_first_purchase_date` and `cumulative_revenue` in the main query; replace `year_month` with `months_since_start` and; only get cohorts 2019 to 2023 and the first 12 months for `months_since_start`.
    - Helps us get long-term revenue growth by cohort (without too much data)
    - Focus on more recent cohorts (within the last 5 years) and excludes cohort 2024 since we only have partial data for that. 

In [35]:
%%sql

WITH cohort_summary AS (
    SELECT
        cc.cohort_year,
        DATE_TRUNC('month', cc.orderdate)::date AS year_month, 
        SUM(cc.total_net_revenue) AS total_revenue
    FROM cleaned_customer cc
    GROUP BY cc.cohort_year, year_month
),

-- Added 
cohort_first_purchase AS (
    SELECT 
        cohort_year,
        MIN(first_purchase_date) AS cohort_first_purchase_date
    FROM cleaned_customer
    GROUP BY cohort_year
),

rolling_ltv AS (
    SELECT
        cs.cohort_year,
        cs.year_month,
        cfp.cohort_first_purchase_date,

        SUM(cs.total_revenue) OVER (
            PARTITION BY cs.cohort_year 
            ORDER BY cs.year_month 
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        ) AS cumulative_revenue,
        DATE_PART('month', age(cs.year_month, cfp.cohort_first_purchase_date)) 
        + DATE_PART('year', age(cs.year_month, cfp.cohort_first_purchase_date)) * 12 
        + 1 AS months_since_start
    FROM cohort_summary cs
    JOIN cohort_first_purchase cfp ON cs.cohort_year = cfp.cohort_year
)

SELECT
    cohort_year,
    months_since_start,
    cumulative_revenue / NULLIF(months_since_start, 0) AS cumulative_avg_ltv
FROM rolling_ltv
WHERE -- Added 
    months_since_start BETWEEN 1 AND 12
    AND cohort_year BETWEEN 2019 AND 2023
ORDER BY 
    cohort_year, 
    year_month;


Running query in 'postgresql://postgres:***@localhost:5432/contoso_100k'

60 rows affected.

cohort_year  months_since_start  cumulative_avg_ltv
0         2019                1.00          2399141.08
1         2019                2.00          2636576.82
2         2019                3.00          2291032.90
3         2019                4.00          1928863.60
4         2019                5.00          1989005.65
5         2019                6.00          1980365.28
6         2019                7.00          1962218.54
7         2019                8.00          1969094.62
8         2019                9.00          1973759.70
9         2019               10.00          1984854.15
10        2019               11.00          1995402.33
11        2019               12.00          2001899.89
12        2020                1.00          1399356.55
13        2020                2.00          1553691.68
14        2020                3.00          1276587.43
15        2020                4.00          1033634.35
16        2020                5.00           987870.20
17        2020                6.00           910436.01
18        2020                7.00           836984.59
19        2020                8.00           770308.60
20        2020                9.00           708955.28
21        2020               10.00           665140.96
22        2020               11.00           628103.22
23        2020               12.00           605479.95
24        2021                1.00           369922.57
25        2021                2.00           492107.40
26        2021                3.00           455828.75
27        2021                4.00           413965.02
28        2021                5.00           535306.60
29        2021                6.00           617784.18
30        2021                7.00           686969.22
31        2021                8.00           741770.75
32        2021                9.00           819835.79
33        2021               10.00           890082.70
34        2021               11.00           954926.18
35        2021               12.00          1052966.49
36        2022                1.00          1877539.46
37        2022                2.00          2205445.74
38        2022                3.00          1935464.79
39        2022                4.00          1695600.69
40        2022                5.00          1835022.40
41        2022                6.00          1979582.72
42        2022                7.00          1978832.23
43        2022                8.00          1982627.92
44        2022                9.00          1984756.84
45        2022               10.00          2007053.40
46        2022               11.00          2003465.36
47        2022               12.00          2034344.84
48        2023                1.00          1604441.44
49        2023                2.00          1700962.83
50        2023                3.00          1445786.95
51        2023                4.00          1213784.57
52        2023                5.00          1211057.93
53        2023                6.00          1211021.63
54        2023                7.00          1198236.34
55        2023                8.00          1202323.04
56        2023                9.00          1199373.65
57        2023               10.00          1178331.82
58        2023               11.00          1188341.65
59        2023               12.00          1195432.08

#### 📊 Key Findings

- 2023 cohort shows highest first-year LTV ($1.2M avg/month)
- 2022 cohort plateaus at $1.1M avg/month after 12 months
- 2019-2021 cohorts show consistent growth pattern:
    - First 3 months: Rapid growth to $800K avg/month
    - Months 4-8: Steady increase to $1M avg/month
    - Months 9-12: Stabilization around $1.1M avg/month

#### 💡 Business Insights
- Cohort Performance:
    - 2023 cohort outperforming previous years by 9% in first-year value
    - Focus on replicating 2023 acquisition strategies
    - Early months crucial for establishing customer value
- Growth Patterns:
    - Critical engagement window identified in months 4-8
    - Implement targeted campaigns during plateau periods
    - Use successful cohort patterns to predict future performance
- Revenue Optimization:
    - Strengthen customer engagement in first 3 months
    - Develop retention strategies for months 4-8 growth period
    - Create stabilization programs for months 9-12

<img src="../8_Project/2.2_cohort_ltv_over_time.png" alt="Cohort LTV Over Time" style="width: 70%; height: auto;">

### 3️⃣ Retention Analysis (Who Hasn’t Purchased Recently?)

#### Overview

- Identify customers at risk of churning.
- Use `ROW_NUMBER()` to track last purchase while capturing revenue insights.

💼 **Example Use Cases:** Optimizes acquisition and retention strategies
- Focus marketing budget on channels producing highest-LTV customers
- Set appropriate customer acquisition costs based on expected lifetime value
- Develop targeted retention programs for highest-potential segments
- Forecast revenue more accurately using cohort performance patterns

#### Query Steps

In [ ]:
%%sql

#### 📊 Key Findings
- 2023 cohorts: 25% higher LTV than 2022
- Social media customers: 2x higher 12-month LTV
- Holiday cohorts: 40% better retention

## Conclusion

Below are the strategic recommendations based on the analysis.

1. **High-Value Focus** ($100K opportunity)
   - Launch premium membership program
   - Deploy churn early warning system
   - Implement proactive service outreach

2. **Acquisition Optimization**
   - Increase social media investment (2x LTV)
   - Optimize seasonal timing
   - Adjust CAC by channel performance

3. **Retention Enhancement**
   - Launch segment-specific reactivation campaigns
   - Create automated upgrade paths
   - Develop targeted loyalty programs